<launch>

    <arg name="sim" default="false" />

    <!-- 
        TODO: Determine if max velocities match the expected range of velocity loop setpoint values. We need to take
        measurements of the max speed the robot can move in x, y, and z directions (local) and set these value 
        accordingly. The units should be meters/sec. Current setup assumes min and max vel are the same.
    /-->
    <arg name="max_control_effort" default="1" />
    <!-- TODO: Determine if orient_max value is correct. Seems like 0.5 was chosen randomly. Currently, this means that
    roll/pitch/yaw control effort can max out at 0.5 This value should likely be 1 to match the x, y, z control effort 
    range. We should test this and replace orient_max with max_control_effort if it works. /-->
    <arg name="orient_max" default="0.5" />

    <arg name="min_loop_freq" value="100" />

    <!-- ROBOT PID Constants -->
    <group unless="$(arg sim)">
	   <group ns="controls/x_pos/controller">
            <param name="Kp" value="1" />
            <param name="Ki" value="0.1" />
            <param name="Kd" value="0.1" />
        </group>
        <group ns="controls/y_pos/controller">
            <param name="Kp" value="1" />
            <param name="Ki" value="0.1" />
            <param name="Kd" value="0.1" />
        </group>
        <group ns="controls/z_pos/controller">
            <!-- Commented values are for camera pods /-->
            <param name="Kp" value="2.32" /> <!-- 25 -->
            <param name="Ki" value="0.3" /> <!-- 0.5 -->
            <param name="Kd" value="3.31" /> <!--0.3 -->
        </group>
        <group ns="controls/roll_pos/controller">
            <param name="Kp" value="0.01" />
            <param name="Ki" value="0.001" />
            <param name="Kd" value="0.001" />
        </group>
        <group ns="controls/pitch_pos/controller">
            <param name="Kp" value="0.01" />
            <param name="Ki" value="0.001" />
            <param name="Kd" value="0.001" />
        </group>
	<group ns="controls/yaw_pos/controller">
            <param name="Kp" value="1" />
            <param name="Ki" value="0.1" />
            <param name="Kd" value="0.1" />
        </group>
    </group>

    <!-- SIMULATION PID Constants -->
    <group if="$(arg sim)">
       <group ns="controls/x_pos/controller">
            <param name="Kp" value="0.7" />
            <param name="Ki" value="0.005" />
            <param name="Kd" value="0.001" />
        </group>
        <group ns="controls/y_pos/controller">
            <param name="Kp" value="0.7" />
            <param name="Ki" value="0.005" />
            <param name="Kd" value="0.001" />
        </group>
        <group ns="controls/z_pos/controller">
            <param name="Kp" value="10" />
            <param name="Ki" value="0.001" />
            <param name="Kd" value="0.001" />
        </group>
        <group ns="controls/roll_pos/controller">
            <param name="Kp" value="0.1" />
            <param name="Ki" value="0.01" />
            <param name="Kd" value="0.01" />
        </group>
        <group ns="controls/pitch_pos/controller">
            <param name="Kp" value="0.1" />
            <param name="Ki" value="0.01" />
            <param name="Kd" value="0.01" />
        </group>
        <group ns="controls/yaw_pos/controller">
            <param name="Kp" value="0.1" />
            <param name="Ki" value="0.01" />
            <param name="Kd" value="0.01" />
        </group>
    </group>

    <!-- Position loop control efforts map to directly to control effort (no more nested loops) /-->  
    <node name="controller" pkg="pid" type="controller" ns="controls/x_pos" output="log" >
        <param name="node_name" value="global_x_pos_pid" />
        <param name="upper_limit" value="$(arg max_control_effort)" /> 
        <param name="lower_limit" value="-$(arg max_control_effort)" />
        <!-- <param name="windup_limit" value="10" /> -->
        <param name="pid_enable_topic" value="/controls/enable/x_pos" />
        <param name="min_loop_frequency" value="$(arg min_loop_freq)" />
        <remap from="control_effort" to="/control_effort/x" />
        <remap from="state" to="/controls/state/pose/x" />
    </node>

    <node name="controller" pkg="pid" type="controller" ns="controls/y_pos" output="log" >
        <param name="node_name" value="global_y_pos_pid" />
        <param name="upper_limit" value="$(arg max_control_effort)" />
        <param name="lower_limit" value="-$(arg max_control_effort)" />
        <!-- <param name="windup_limit" value="10" /> -->
        <param name="pid_enable_topic" value="/controls/enable/y_pos" />
        <param name="min_loop_frequency" value="$(arg min_loop_freq)" />
        <remap from="control_effort" to="/control_effort/y" />
        <remap from="state" to="/controls/state/pose/y" />
    </node>

    <node name="controller" pkg="pid" type="controller" ns="controls/z_pos" output="log" >
        <param name="node_name" value="global_z_pos_pid" />
        <param name="upper_limit" value="$(arg max_control_effort)" />
        <param name="lower_limit" value="-$(arg max_control_effort)" />
        <!-- <param name="windup_limit" value="10" /> -->
        <param name="pid_enable_topic" value="/controls/enable/z_pos" />
        <param name="min_loop_frequency" value="$(arg min_loop_freq)" />
        <remap from="control_effort" to="/control_effort/z" />
        <remap from="state" to="/controls/state/pose/z" />
    </node>

    <node name="controller" pkg="pid" type="controller" ns="controls/roll_pos" output="log" >
        <param name="node_name" value="global_roll_pos_pid" />
        <param name="upper_limit" value="$(arg orient_max)" />
        <param name="lower_limit" value="-$(arg orient_max)" />
        <!-- <param name="windup_limit" value="10" /> -->
        <param name="pid_enable_topic" value="/controls/enable/roll_pos" />
        <param name="min_loop_frequency" value="$(arg min_loop_freq)" />
        <remap from="control_effort" to="/control_effort/roll" />
        <remap from="state" to="/controls/state/pose/roll" />
    </node>

    <node name="controller" pkg="pid" type="controller" ns="controls/pitch_pos" output="log" >
        <param name="node_name" value="global_pitch_pos_pid" />
        <param name="upper_limit" value="$(arg orient_max)" />
        <param name="lower_limit" value="-$(arg orient_max)" />
        <!-- <param name="windup_limit" value="10" /> -->
        <param name="pid_enable_topic" value="/controls/enable/pitch_pos" />
        <param name="min_loop_frequency" value="$(arg min_loop_freq)" />
        <remap from="control_effort" to="/control_effort/pitch" />
        <remap from="state" to="/controls/state/pose/pitch" />
    </node>

    <node name="controller" pkg="pid" type="controller" ns="controls/yaw_pos" output="log" >
        <param name="node_name" value="global_yaw_pos_pid" />
        <param name="upper_limit" value="$(arg orient_max)" />
        <param name="lower_limit" value="-$(arg orient_max)" />
        <!-- <param name="windup_limit" value="10" /> -->
        <param name="pid_enable_topic" value="/controls/enable/yaw_pos" />
        <param name="min_loop_frequency" value="$(arg min_loop_freq)" />
        <remap from="control_effort" to="/control_effort/yaw" />
        <remap from="state" to="/controls/state/pose/yaw" />
    </node>

</launch>
